drivers/drivers_4in2.py

```diff
@@ -421,4 +375,6 @@ def draw(self, x, y, image):
             self.display_partial(y, x, y + image.height, x + image.width)
         else:
             self.set_frame_buffer(0, 0, image)
+            # this one is much faster!!!
+            # self.display_partial(y, x, y + image.height, x + image.width)
             self.display_full()
```

Comment:
I think `display_full` is not meant for continuous refresh, as it spends most time in doing full black/white scrubs, I would advocate for simply using `display_partial`.

Reply:
This depends on the use case, not all of them require fast updates (consider a wall calendar for example) and instead prefer that the image is as crisp as possible and maintenance-free. Only doing partial updates will quite soon make the display ghosty until a full refresh is done. The VNC mode by default does a full refresh every 50 updates to mitigate the problem, but perhaps there could be some smarter heuristic too.

In any case Waveshare's wiki (https://www.waveshare.com/wiki/4.2inch_e-Paper_Module) has these kinds of mentions:

> Note that you cannot use Partial refresh all the time, you should full refresh e-paper regularly, otherwise, ghost problem will get worse even damage.

Not sure how long it would take for actual damage to be caused (probably depends very much on the display module if it even happens in practice), but we should just have a smart balance between full and partial refreshes and options to configure it as needed.

Reply:
so in conclusion: We use `display_partial(0,0)` here and do a full refresh every once in a while?

Reply:
I think if the user wants full refreshes (as opposed to partial ones), we should then actually do a full refresh every time. In cases where the update interval might be set to hours, the degraded appearance would be visible for quite some time before the "once in a while" happens. If partial refresh is supported, then perhaps the safest option would be to prefer them but *by default* also do the periodical full refresh to keep the image quality reasonable, while allowing to configure the interval (or disable it altogether if that is desired).

This way those who want nothing but full refreshes get them, those who want nothing but partial refreshes can disable the full refresh, and by default it will do partial refresh with the periodic full ones as per Waveshare's recommendations and hopefully doesn't break anything.

The partial refresh code may not always be very "official" either as in the case of the 4.2" which lacks such examples but technically does support it. I've gotten mine in a semi-"bricked" state a couple of times when experimenting with it, so I'd like to minimize any risk to other people's devices. Their display just might be from a batch with a slightly different internal configuration or components and doesn't work the same.

My 2.13" is one example of such - it can do partial refresh but *doesn't do full refresh* as documented (and never has) - that's why I implemented the scrub routine, but after contacting Waveshare about it, they said that my unit is probably broken.

Reply:
sounds reasonable to me!